File: src/main/java/se/viati/stockholm/services/SlackService.java
package se.viati.stockholm.services;

import net.gpedro.integrations.slack.SlackApi;
import net.gpedro.integrations.slack.SlackMessage;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.safety.Whitelist;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import se.viati.stockholm.services.domain.Mail;

import javax.annotation.PostConstruct;

@Service
public class SlackService {

    @Value("${slack.webhook}")
    private String slackWebHook;
    @Value("${slack.channel.full}")
    private String channelFull;
    @Value("${slack.channel.titles}")
    private String channelTitles;

    private SlackApi slackApi;

    private static final Logger logger = LoggerFactory.getLogger(SlackService.class);

    @PostConstruct
    public void init() {
        slackApi = new SlackApi(slackWebHook);
    }

    public void postToSlack(Mail mail) {
        // Send simple message with custom name
        logger.info("Posting to Slack: " + mail.subject);
        final SlackMessage messageWithBody = new SlackMessage(channelFull,
                null, "*" + mail.subject + "*" +
                "\n" + getHtmlAsStringWithLineBreaks(mail.body));
        slackApi.call(messageWithBody);

        final SlackMessage messageWithOnlySubject = new SlackMessage(channelTitles,
                null, "*" + mail.subject + "*");
        slackApi.call(messageWithOnlySubject);
    }

    private String getHtmlAsStringWithLineBreaks(String html) {
        final Document bodyDocument = Jsoup.parse(html);
        bodyDocument.outputSettings(new Document.OutputSettings().prettyPrint(false));//makes html() preserve linebreaks and spacing
        bodyDocument.select("br").append("\\n");
        bodyDocument.select("p").prepend("\\n\\n");
        String s = bodyDocument.html()
                .replaceAll("\\\\n", "\n");
        final String htmlAsText = Jsoup.clean(s, "", Whitelist.none(), new Document.OutputSettings().prettyPrint(false));
        return htmlAsText
                .replace("&nbsp;", "")
                .replace("&amp;", "&")
                .replace("\n\n\n\n", "\n")
                .replace("\n\n\n", "\n")
                .replace("\n\n", "\n");
    }

}

File: src/main/java/se/viati/stockholm/services/SlackService.kt
package se.viati.stockholm.services

import net.gpedro.integrations.slack.SlackApi
import net.gpedro.integrations.slack.SlackMessage
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import se.viati.stockholm.services.domain.Assignment

import javax.annotation.PostConstruct

@Service
class SlackService {

  @Value("\${slack.enabled}")
  private val slackEnabled: Boolean = false
  @Value("\${slack.webhook}")
  private val slackWebHook: String? = null
  @Value("\${slack.channel.full}")
  private val channelFull: String? = null
  @Value("\${slack.channel.titles}")
  private val channelTitles: String? = null

  private lateinit var slackApi: SlackApi

  private val logger = LoggerFactory.getLogger(this::class.java)

  @PostConstruct
  fun init() {
    slackApi = SlackApi(slackWebHook!!)
  }

  fun postToSlack(assignment: Assignment) {
    val assignmentTitle = "*${assignment.title}* ${assignment.source?.let { "[$it]" } ?: ""}"

    logger.info("${if (slackEnabled) "Posting" else "Not posting (disabled)"} to Slack: $assignmentTitle, mails: ${assignment.originatingMailIds.size}")

    if (slackEnabled) {
      val messageWithBody = SlackMessage(channelFull,
          null, "$assignmentTitle\n${assignment.description}")
      slackApi.call(messageWithBody)

      val messageWithOnlyTitle = SlackMessage(channelTitles,
          null, assignmentTitle)
      slackApi.call(messageWithOnlyTitle)
    }
  }
}
